src/AddAtomicMutex.cpp

```diff
@@ -0,0 +1,266 @@
+
+    FindAtomicLetBindings(const set<string> &store_names)
+        : store_names(store_names) {}
+
+    void visit(const Let *op) override {
+        include(op->value);
+        let_bindings.push(op->name, op->value);
+        include(op->body);
+        let_bindings.pop(op->name);
+    }
+
+    void visit(const LetStmt *op) override {
+        include(op->value);
+        let_bindings.push(op->name, op->value);
+        include(op->body);
+        let_bindings.pop(op->name);
+    }
+
+    void visit(const Variable *op) override {
+        if (inside_store) {
+            if (let_bindings.contains(op->name)) {
+                Expr e = let_bindings.get(op->name);
+                FindLoad finder;
+                if (finder.find_load(e, store_names)) {
+                    found = true;
+                }
+            }
+        }
+    }
+
+    void visit(const Store *op) override {
+        include(op->predicate);
+        inside_store = true;
+        include(op->value);
+        inside_store = false;
+        include(op->index);
+    }
+
+    bool inside_store;
+    set<string> store_names;
+    Scope<Expr> let_bindings;
+    bool found = false;
+};
+
+class CollectStoreNames : public IRGraphVisitor {
+public:
+    using IRGraphVisitor::visit;
+    
+    void visit(const Store *op) override {
+        include(op->predicate);
+        include(op->value);
+        include(op->index);
+        store_names.insert(op->name);
+    }
+
+    set<string> store_names;
+};
+
+class RemoveUnnecessaryMutexUse : public IRMutator {
+public:
+    using IRMutator::visit;
+
+    Stmt visit(const Atomic *op) override {
+        // Collect the names of all Store nodes inside.
+        CollectStoreNames collector;
+        op->body.accept(&collector);
+        // Search for let bindings that access the providers.
+        FindAtomicLetBindings finder(collector.store_names);
+        op->body.accept(&finder);
+        if (finder.found) {
+            // Can't remove mutex lock. Leave the Stmt as is.
+            return IRMutator::visit(op);
+        } else {
+            remove_mutex_lock_names.insert(op->mutex_name);
+            Stmt body = mutate(op->body);
+            return Atomic::make(op->producer_name,
+                                "",
+                                {},
+                                op->tuple_size,
+                                op->dimensions,
+                                std::move(body));
+        }
+    }
+
+    std::set<string> remove_mutex_lock_names;
+};
+
+class FindAtomicMutexUsage : public IRGraphVisitor {
+public:
+    FindAtomicMutexUsage(const std::string &producer_name)
+        : producer_name(producer_name) {}
+
+    using IRGraphVisitor::visit;
+
+    void visit(const Atomic *op) override {
+        for (Expr i : op->mutex_indices) {
+            include(i);
+        }
+        include(op->body);
+        if (op->producer_name == producer_name &&
+                op->mutex_name != "") {
```

Comment:
Use string::empty() instead of comparing it to "" here and elsewhere

Reply:
Done.